pkg/agent/controller/traceflow/traceflow_controller.go

```diff
@@ -0,0 +1,332 @@
+				} else {
+					klog.Warningf("Invalid data plane tag %d for Traceflow %s", tf.Status.DataplaneTag, tf.Name)
+				}
+			default:
+				c.deallocateTag(&tf)
+			}
+		}
+		return false, nil
+	}, stopCh)
+	<-stopCh
+}
+
+// worker is a long-running function that will continually call the processTraceflowItem function
+// in order to read and process a message on the workqueue.
+func (c *Controller) worker() {
+	for c.processTraceflowItem() {
+	}
+}
+
+// processTraceflowItem processes an item in the "traceflow" work queue, by calling syncTraceflow
+// after casting the item to a string (Traceflow name). If syncTraceflow returns an error, this
+// function handles it by updating phase to "ERROR". If syncTraceflow is successful, the Traceflow
+// is removed from the queue until we get notified of a new change. This function returns false if
+// and only if the work queue was shutdown (no more items will be processed).
+func (c *Controller) processTraceflowItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	// We call Done here so the workqueue knows we have finished processing this item. We also
+	// must remember to call Forget if we do not want this work item being re-queued. For
+	// example, we do not call Forget if a transient error occurs, instead the item is put back
+	// on the workqueue and attempted again after a back-off period.
+	defer c.queue.Done(obj)
+
+	// We expect strings (Node name) to come off the workqueue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen: enqueueTraceflow only enqueues strings.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncTraceflow(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		klog.Errorf("Error syncing Traceflow %s, Aborting. Error: %v", key, err)
+		c.queue.Forget(key)
+	}
+	time.Sleep(time.Duration(2) * time.Second)
+	return true
+}
+
+// syncTraceflow
+func (c *Controller) syncTraceflow(traceflowName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Traceflow for %s. (%v)", traceflowName, time.Since(startTime))
+	}()
+
+	tf, err := c.traceflowClient.OpsV1alpha1().Traceflows().Get(traceflowName, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	c.updateTagCache(tf)
+	return c.startTraceflow(tf)
+}
+
+// startTraceflow
+func (c *Controller) startTraceflow(tf *opsv1alpha1.Traceflow) error {
+	// Deploy flow entries for traceflow
+	err := c.deployFlowEntries(tf)
+	defer func() {
+		if err != nil {
+			c.errorTraceflowCRD(tf)
+		}
+	}()
+	if err != nil {
+		return err
+	}
+
+	// Inject packet if this Node is sender
+	time.Sleep(time.Duration(2) * time.Second)
```

Comment:
why sleeping here? If it's waiting for some condition, Is 2 seconds reliable?

Reply:
Different Nodes may not start Traceflow at same time (e.g. heavy load on agent or slow response on apiserver), thus we must wait a short period before inject packet.
Do you have a better suggestion for this?

Reply:
Added constant and comments for this.
	// Seconds delay before inject packet into OVS. The time of different nodes may not be completely
	// synchronized, which require a delay before inject packet
	injectPacketDelay = 2